Simplify and clarify the memory terminology article

captainbrosset opened the pull request:
Rendered articles for review:
* Memory terminology: https://review.learn.microsoft.com/microsoft-edge/devtools-guide-chromium/memory-problems/memory-101?branch=pr-en-us-3033 [[Before/Live](https://learn.microsoft.com/microsoft-edge/devtools-guide-chromium/memory-problems/memory-101)]
* Record heap snapshots: https://review.learn.microsoft.com/microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots?branch=pr-en-us-3033 [[Before/Live](https://learn.microsoft.com/microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots)]

These articles are a bit hard to understand, sometimes give information that's not really actionable for web developers, and could use an update.

AB#48649923

captainbrosset commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
+## Garbage-collection roots
 
-_Garbage-collection roots_ are made up of _handles_ that are created (as either local or global) when making a reference from native code to a JavaScript object outside of the V8 VM.  All such handles can be found within a heap snapshot under **GC roots** > **Handle scope** and **GC roots** > **Global handles**.  Describing the handles in this documentation without diving into details of the brow ...
+_Garbage-collection roots_, or GC roots, are created by the browser when a reference from the browser's native code to a JavaScript object outside of the V8 virtual machine is made. These references are called _handles_. You can find these handles in the **Memory** tool under **GC roots** > **Handle scope** and **GC roots** > **Global handles**.<!-- Can't find this in the Memory tool. Where is it ...
```
@sethbrenith do you know if GC roots are still visible in the Memory tool, and if so, where?

sethbrenith replied:
They're mostly hidden most of the time, and that's on purpose. By default, heap snapshots lie a little bit and pretend that any attached global objects (`Window`) are the roots. It's possible to see the term "GC roots" in the Containment view, but relatively little memory is attributed to that category.

However, the _idea_ of roots is somewhat important for understanding how a garbage collector works, which is the topic of a few paragraphs down from here.

captainbrosset replied:
Thanks, removing the note about **Global handles** in DevTools, but keeping this intro about GC roots.

captainbrosset commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
+There are many internal GC roots, most of which aren't interesting for web developers. From the webpage's standpoint, the following types of GC roots exist:
 
-*  Window global object (in each iframe).  In the heap snapshots, the `distance` field indicates the number of property references on the shortest retaining path from the window.
+*  Window global object (in each iframe). In a heap snapshot displayed in the **Memory** tool, the `distance` field for a window global object indicates the number of property references that on the shortest retaining path from the window.<!-- If this is useful, this needs clarification: what window? The top-level one? Does retaining path mean the path through the graph from the iframe window to  ...
```
@sethbrenith this sentence is hard to parse. I'm wondering if it's still useful, and if it is, would you have a suggestion for making it clearer?

sethbrenith replied:
Agreed: this sentence seems low-value and sort of irrelevant to the primary topic of this section, which is describing the roots.

captainbrosset replied:
Removed the sentence about the distance field.

captainbrosset commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
-![Node #7 is dominated by node #3 and dominates nodes #8, #9, and #10](./memory-101-images/dominators-02.png)
- 
-Node `#8` is dominated by node `#7` and doesn't dominate any nodes:
+If almost all the objects of the same type are at the same distance, and a few are at a bigger distance, that's something worth investigating.<!-- WHY? -->
```
@sethbrenith My understanding is that the `distance` column in the Memory tool isn't super actionable for web developers. This sentence has been in this article since the beginning, and I don't really know what it means. Why would looking at this pattern help web devs?

captainbrosset replied:
Thanks. Removing the sentence, but adding a sentence about the Retainers section default sorting order.

captainbrosset commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -192,46 +173,47 @@ A typical JavaScript object can be one of two array types:
-Each _native objects group_ is made up of objects that hold mutual references to each other.  Consider, for example, a DOM subtree where every node has a link to the relative parent and links to the next child and next sibling, thus forming a connected graph.
+<!-- ------------------------------ -->
+#### Object groups
+<!-- This doesn't seem very actionable to users. We describe groups, but don't say where they appear in the tool and what's interesting about them. We then talk about native objects, seemingly out of the blue. -->
```
@sethbrenith See my HTML comment here. Are object groups still a thing in V8, and do they play a role in memory consumption that web developers should know about?

sethbrenith replied:
I don't think any of this is relevant anymore; let's get rid of the section.

captainbrosset commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -192,46 +173,47 @@ A typical JavaScript object can be one of two array types:
 <!-- ====================================================================== -->
 ## Cycles
-
+<!-- This needs a concrete example, not sure what earlier and later mean. And why is removing the earlier instance important? -->
```
@sethbrenith any help on this would be great!

sethbrenith replied:
Hmm, I agree that this section is pretty confusing, and I'm not quite sure how to make it better. The simplest advice would be: when trying to remove an object from the heap, it's generally most helpful to focus on the shortest retaining path, which is expanded by default when you select the object.

The Retainers pane shows all objects which point to the selected object. Following along edges backwards through the graph, you might end up revisiting a node you've already seen, which is not super useful, so Jose added a filter to hide such nodes in the Retainers view. However, I'm not sure that this section fits conceptually with the rest of this article: the article seems to be high-level ideas to get someone started on memory analysis, and this section is a description of a specific feature to solve a specific problem. It might be best to just remove this section.

captainbrosset replied:
Thanks. I'll consider moving these notes to our Memory tool reference article instead then.

sethbrenith commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
+## JavaScript heap and renderer memory
 
-Typical JavaScript objects have some memory reserved for their description and for storing immediate values.  Usually, only arrays and strings can have a significant shallow size.  However, strings and external arrays often have their main storage in renderer memory, exposing only a small wrapper object on the JavaScript heap.
+The JavaScript heap is a memory region in the browser process where all the JavaScript objects live. The JavaScript heap is what the **Memory** tool displays when you take a heap snapshot. The JavaScript heap is also called the _V8 heap_ (after the V8 JavaScript engine that powers Microsoft Edge).
```
> "The JavaScript heap is what the **Memory** tool displays when you take a heap snapshot.

This used to be true, but over the last few years, upstream folks have put a lot of effort into making the V8 and Blink garbage collectors work together better. Now the heap snapshot shows both V8 memory (which contains JS and WebAssembly), and Blink objects allocated in the Blink memory region. I'm not sure how to say that in a way that is both true and easy to read.

sethbrenith commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
+*  Window global object (in each iframe). In a heap snapshot displayed in the **Memory** tool, the `distance` field for a window global object indicates the number of property references that on the shortest retaining path from the window.<!-- If this is useful, this needs clarification: what window? The top-level one? Does retaining path mean the path through the graph from the iframe window to  ...
 
-*  The document DOM tree, consisting of all native DOM nodes that are reachable by traversing the document.  Not all of the nodes have JavaScript wrappers, but if a node has a wrapper, the node is alive while the document is alive.
+*  The document DOM tree, consisting of all native DOM nodes that are reachable by traversing the document. Not all of the nodes have JavaScript wrappers, but if a node has a wrapper, the node is alive while the document is alive.
```
I don't think that this is true anymore. I think that the Memory tool currently treats only `Window` instances as roots (both the top-level one and any same-domain iframes). When I select any DOM nodes in the Memory tool, I see a retaining path describing how the DOM node is kept alive by a Window.

sethbrenith commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
 
-<!-- ====================================================================== -->
-## Dominators
+The shallow size of an object is usually small because JavaScript objects often just store their own descriptions only in memory. Arrays and strings are exceptions to this and can have a significant shallow size. Most JavaScript objects store their values elsewhere in the renderer memory, and only expose a small wrapper object on the JavaScript heap.
```
> Most JavaScript objects store their values elsewhere in the renderer memory, and only expose a small wrapper object on the JavaScript heap.

By the way, I'm working on changing this so that shallow sizes will be more useful. The experimental setting will land upstream soon (probably tomorrow). Work is tracked [here](https://bugs.chromium.org/p/chromium/issues/detail?id=1517057).

captainbrosset replied:
Cool, that sounds very useful. Once the new setting is in Canary, we should document it too.

sethbrenith commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
-*  Node 3 dominates node 5.
-*  Node 5 dominates node 8.
-*  Node 6 dominates node 7.
+The _retained size_ is the size of memory that is implicitly held by an object and that is freed once the object is deleted along with all the dependent objects that were made unreachable from GC roots.
```
Two comments here:

1. I would say "would be freed" rather than "is freed", since we haven't yet freed the object (it's in the snapshot, after all)
2. Christy changed what this column means. I'm not really sure how to describe the meaning of the new value; you might want to consult with her. https://dev.azure.com/microsoft/Edge/_git/chromium.devtools-frontend/pullrequest/9283784

captainbrosset replied:
@christyycchen any proposal for a revised wording here, that would match the new retained size column counting logic?

christyycchen replied:
Something around 
"is the size of memory that is implicitly held by an object and that could be freed once the object and other existing retainer is deleted along with all the dependent objects that were made unreachable from GC roots." ? 🤔
Maybe add some notes explaining that when a child node is retained by multiple nodes, the size of that child node is contributed to the parent node that has the shorted path to the GC root. Some graphs could help explaining this.

sethbrenith commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
+**Native objects** - Anything not in the JavaScript heap is called a _native object_. A native object, in contrast to a heap object, isn't managed by the V8 garbage collector throughout its lifetime, and can only be accessed from JavaScript by using its JavaScript wrapper object.
 
-A **cons string** (concatenation string) is an object that consists of pairs of strings that are stored and then joined, and is a result of concatenation.  The joining of the **cons string** contents occurs only as needed.  For example, when a substring of a joined string needs to be constructed.
+A **cons string** (concatenation string) is an object that consists of pairs of strings that are stored and then joined, and is a result of concatenation. The joining of the **cons string** contents occurs only as needed. For example, when a substring of a joined string needs to be constructed.
```
Nit: in the DevTools, these are called "concatenat*ed* string", not "concatenat*ion* string".

mikehoffms replied:
```suggestion
A _cons string_ (concatenated string) is an object that consists of pairs of strings that are stored and then joined, and is a result of concatenation. The joining of the **cons string** contents occurs only as needed, such as when a substring of a joined string needs to be constructed.
```

sethbrenith commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -192,46 +173,47 @@ A typical JavaScript object can be one of two array types:
-**Map** is an object that describes both the kind of object it is and the layout.  For example, maps are used to describe implicit object hierarchies for [fast property access](https://v8.dev/blog/fast-properties).
-
-### Object groups
+**Map** is an object that describes both the kind of object it is and the layout. For example, maps are used to describe implicit object hierarchies for [fast property access](https://v8.dev/blog/fast-properties).
```
In the DevTools, this is called `system / Map`. `Map` is used to refer to JavaScript [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) objects.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
+There are many internal GC roots, most of which aren't interesting for web developers. From the webpage's standpoint, the following types of GC roots exist:
 
-*  Window global object (in each iframe).  In the heap snapshots, the `distance` field indicates the number of property references on the shortest retaining path from the window.
+*  Window global object (in each iframe). In a heap snapshot displayed in the **Memory** tool, the `distance` field for a window global object indicates the number of property references that on the shortest retaining path from the window.<!-- If this is useful, this needs clarification: what window? The top-level one? Does retaining path mean the path through the graph from the iframe window to  ...
```
```suggestion
*  Window global object (in each iframe). In a heap snapshot displayed in the **Memory** tool, the `distance` field for a window global object indicates the number of property references that are on the shortest retaining path from the window.<!-- If this is useful, this needs clarification: what window? The top-level one? Does retaining path mean the path through the graph from the iframe window to the top-level window? -->
```
* that are

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
+*  The document DOM tree, consisting of all native DOM nodes that are reachable by traversing the document. Not all of the nodes have JavaScript wrappers, but if a node has a wrapper, the node is alive while the document is alive.
 
-*  Sometimes objects are retained by the debug context in the **Sources** tool and the **Console**, such as after Console evaluation.  Create heap snapshots with a cleared **Console** tool and no active breakpoints in the debugger in the **Sources** tool.
+*  Sometimes objects are retained by the debugging context set by the **Sources** or the **Console** tool, for example when evaluating a JavaScript expression in the **Console** tool. To remove these objects from the **Memory** tool, clear the **Console** tool and deactivate breakpoints in the **Sources** tool before recording a heap snapshot.
```
```suggestion
*  Sometimes objects are retained by the debugging context that's set by the **Sources** or the **Console** tool, for example when evaluating a JavaScript expression in the **Console** tool. To remove these objects from the **Memory** tool, clear the **Console** tool and deactivate breakpoints in the **Sources** tool before recording a heap snapshot.
```
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/t/that-vs-which - "Include _that_ even if the sentence is clear without it. It helps to clarify the sentence structure."

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
+* **Distance**
 
-*  _Nodes_ (or _objects_) are labelled using the name of the _constructor_ function that was used to build them.
+The numbers that are shown in the **Shallow size** and **Retained size** columns are the number of bytes.
```
```suggestion
The numbers in the **Shallow size** and **Retained size** columns are the number of bytes.
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
-Learn [how to record a profile using the Heap Profiler](heap-snapshots.md).  In the following figure, some of the notable things in the Heap Snapshot recording in the [Memory](heap-snapshots.md) tool include **Distance**, which is the distance from the garbage-collection root.  If almost all the objects of the same type are at the same distance, and a few are at a bigger distance, that's somethin ...
-
-Distance from root:
+![Shallow and Retained Size](./memory-101-images/shallow-retained.png)
```
```suggestion
![Shallow and Retained size](./memory-101-images/shallow-retained.png)
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
 
-<!-- ====================================================================== -->
-## Dominators
+The shallow size of an object is usually small because JavaScript objects often just store their own descriptions only in memory. Arrays and strings are exceptions to this and can have a significant shallow size. Most JavaScript objects store their values elsewhere in the renderer memory, and only expose a small wrapper object on the JavaScript heap.
```
```suggestion
The shallow size of an object is usually small because JavaScript objects often store their own descriptions only in memory, rather than in __. Arrays and strings are exceptions to this and can have a significant shallow size. Most JavaScript objects store their values elsewhere in the renderer memory, and only expose a small wrapper object on the JavaScript heap.
```
need to clarify the negative - stored in memory, as opposed to what - renderer memory?

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
+The shallow size of an object is usually small because JavaScript objects often just store their own descriptions only in memory. Arrays and strings are exceptions to this and can have a significant shallow size. Most JavaScript objects store their values elsewhere in the renderer memory, and only expose a small wrapper object on the JavaScript heap.
 
-Dominator objects are comprised of a tree structure, because each object has exactly one dominator.  A dominator of an object might lack direct references to an object that it dominates.  That is, dominator's tree is not a spanning tree of the graph.
+Nevertheless, even a small object can hold a large amount of memory indirectly, by preventing other objects from being disposed of by the automatic garbage collection process.
```
```suggestion
Nevertheless, even a small object can hold a large amount of memory indirectly, by preventing other objects from being disposed of by the garbage collection process.
```
* automatic - GC is inherently automatic

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
+The _retained size_ is the size of memory that is implicitly held by an object and that is freed once the object is deleted along with all the dependent objects that were made unreachable from GC roots.
 
-![Dominator tree structure](./memory-101-images/dominatorsspanning.png)
+Said another way, the retained size of an object is the amount of memory that would be regained if the object and all of its dependent objects were removed from the memory graph.
```
```suggestion
That is, the retained size of an object is the amount of memory that would be regained if the object and all of its dependent objects were removed from the memory graph.
```
more standard phrase imo

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
 ## V8 specifics
 
-When profiling memory, it is helpful to understand why heap snapshots look a certain way.  This section describes some memory-related topics specifically corresponding to the _V8 JavaScript virtual machine_ (abbreviated here as _V8 VM_, or just _VM_).
+When profiling memory, it is helpful to understand why heap snapshots look a certain way. This section describes some memory-related topics specifically corresponding to the _V8 JavaScript virtual machine_ (abbreviated here as _V8 VM_, or just _VM_).
```
```suggestion
When profiling memory, it's helpful to understand why heap snapshots look a certain way. This section describes some memory-related topics specifically corresponding to the _V8 JavaScript virtual machine_ (abbreviated here as _V8 VM_, or just _VM_).
```
brand voice is contractions, except where emphasis is needed

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
 *  The **VM heap**.
 
-*  Externally in the **renderer's memory**.  A _wrapper object_ is created and used for accessing external storage where, for example, script sources and other content that is received from the Web is stored, rather than copied onto the VM heap.
+*  Externally in the **renderer's memory**. A _wrapper object_ is created and used for accessing external storage where, for example, script sources and other content that is received from the Web is stored, rather than copied onto the VM heap.
```
```suggestion
*  Externally in the **renderer's memory**. A _wrapper object_ is created and used for accessing external storage where, for example, script sources and other content that's received from the Web is stored, rather than copied onto the VM heap.
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,165 +23,146 @@ ms.date: 07/21/2023
+A **cons string** (concatenation string) is an object that consists of pairs of strings that are stored and then joined, and is a result of concatenation. The joining of the **cons string** contents occurs only as needed. For example, when a substring of a joined string needs to be constructed.
 
-For example, if you concatenate `a` and `b`, you get a string `(a, b)` which represents the result of concatenation, and is a cons string.  If you later concatenated `d` with that result, you get another cons string: `((a, b, d)`.
+For example, if you concatenate `a` and `b`, you get a string `(a, b)` which represents the result of concatenation, and is a cons string. If you later concatenated `d` with that result, you get another cons string: `((a, b, d)`.
```
```suggestion
For example, if you concatenate `a` and `b`, you get a string `(a, b)` which represents the result of concatenation, and is a cons string. If you later concatenated `c` with that result, you get another cons string: `((a, b, c)`.
```
* c - doesn't raise questions, like "d".  Is there some reason "d" was used instead of "c"?

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101-images/thinkgraph.png:
too jumbo, makes harder to read.  in Paint, Resize eg to 67%.  Keep open, in case undo > redo w/ different %.

rendered section:
https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/memory-problems/memory-101?branch=pr-en-us-3033#the-memory-graph

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,216 +23,166 @@ ms.date: 07/21/2023
+* **Concatenated strings**: strings that are stored and then joined together, as a result of a string concatenation in JavaScript, are stored as _concatenated strings_ in V8. The joining of the string contents occurs only as needed, such as when a substring of a joined string needs to be constructed.
 
-In the **Filter edges** dropdown menu, **Hide cycled** is selected, so the cycled node is not displayed:
+  For example, if you concatenate `a` and `b`, you get a concatenated string `(a, b)` which represents the result of concatenation. If you later concatenate `c` with that result, you get another concatenated string: `((a, b, c)`.
```
```suggestion
  For example, if you concatenate `a` and `b`, you get a concatenated string `(a, b)` which represents the result of concatenation. If you later concatenate `c` with that result, you get another concatenated string: `(a, b, c)`.
```
@captainbrosset tech review this suggestion.  you might want eg (a b (c))

captainbrosset replied:
It should be `((a, b), c)`. This was a typo. Fiuxing.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,216 +23,166 @@ ms.date: 07/21/2023
+  For example, if you concatenate `a` and `b`, you get a concatenated string `(a, b)` which represents the result of concatenation. If you later concatenate `c` with that result, you get another concatenated string: `((a, b, c)`.
 
-![In the 'Filter edges' dropdown menu, 'Hide cycled' is selected](memory-101-images/filters-retainers-memory-tool-hide-cycled.png)
+* **Arrays**: an _array_ is an object that has numeric keys. Arrays are used extensively in the V8 VM for storing large amounts of data. Sets of key-value pairs that are used like dictionaries are implemented as **arrays**.
```
```suggestion
* **Arrays**: an _array_ is an object that has numeric keys. Arrays are used extensively in the V8 VM for storing large amounts of data. Sets of key-value pairs that are used like dictionaries are implemented as arrays.
```
is bolding appropriate?  'arrays' seems like a plain regular word, after the introduction highlights it.  Same w/ next sentence.  If "arrays" is a lang keyword/type, it should be backticks, not bold.

mikehoffms replied:
removed bold 2x, seems just as clear.

sethbrenith commented:
Looking great, thanks! I have a couple of comments on parts that you didn't change in the article "Record heap snapshots using the Memory tool", which you're welcome to add to this pull request or leave alone for now.

* The section named "Look up color coding" doesn't render with color or describe any colors.
* The section "Constructor (group) entries in the heap profiler" has some inaccuracies. I've never seen "(roots)" or "(global property)" in that list, and the description for "(array)" is incorrect. I also recently [got rid of the category named "(closure)" and "(regexp)"](https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/5193663), though the current Stable and Beta versions still show those names. Here's a [list I wrote describing the weird names that show up in the Summary](https://github.com/sethbrenith/sethbrenith.github.io/blob/main/heap-snapshot-names.md). I also shared that list with Sofia (the upstream documentation writer), who said she'll update the upstream docs soon.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,216 +23,166 @@ ms.date: 07/21/2023
 
 In JavaScript, there are three primitive types:
 
 *  Numbers (such as `3.14159...`).
```
@captainbrosset minor note: the string or value "3.14159..." with dots isn't a number type; it's not an actual example of a number value AFAIK.  removed dots.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,216 +23,166 @@ ms.date: 07/21/2023
 # Memory terminology
 
-This article describes common terms used in memory analysis, and is applicable to various memory profiling tools for different languages.
+This article describes common terms used in the **Memory** tool, a tool for investigating memory issues, but is applicable to memory analysis in general and to various memory profiling tools for different languages. If you've worked with memory profilers for languages such as Java or .NET, this article may be a refresher.
```
```suggestion
The following terms are used in the **Memory** tool, which is for investigating memory issues.  These memory terms are applicable to memory analysis in general, as well as memory profiling tools for languages such as Java or .NET.
```
@captainbrosset - did this interesting rewrite of important Paragraph 1 fyi.
* Talk in terms of tech, not docs.  Keep focus on tech, not docs.
* Break up long sentence, as Acrolinx probably says.
* Uncomfortable talking about the reader's level of knowledge; not necessary.  Keep focus on tech, not speculating about user's level of knowledge.

captainbrosset commented:
> * The section named "Look up color coding" doesn't render with color or describe any colors.

Removed this entire section which was just very confusing to me. I couldn't recognize anything from the docs when looking at an actual snapshot.

> * The section "Constructor (group) entries in the heap profiler" has some inaccuracies. I've never seen "(roots)" or "(global property)" in that list, and the description for "(array)" is incorrect. I also recently [got rid of the category named "(closure)" and "(regexp)"](https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/5193663), though the current Stable and Beta versions still show those names. Here's a [list I wrote describing the weird names that show up in the Summary](https://github.com/sethbrenith/sethbrenith.github.io/blob/main/heap-snapshot-names.md). I also shared that list with Sofia (the upstream documentation writer), who said she'll update the upstream docs soon.

I've changed the entire table based on your excellent summary.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,222 +23,175 @@ ms.date: 07/21/2023
+* **Distance**
 
-*  _Edges_ are labelled using the names of _properties_.
+The numbers in the **Shallow size** and **Retained size** columns are the number of bytes.
```
```suggestion
The numbers in the **Shallow size** and **Retained size** columns are the number of bytes:
```
check this suggestion by looking at flow in rendered article section: 
pending

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,222 +23,175 @@ ms.date: 07/21/2023
+When working with the **Memory** tool, you will likely find yourself looking at the following columns of information:
 
-*  _Nodes_ (or _objects_) are labelled using the name of the _constructor_ function that was used to build them.
+* **Shallow Size**
+* **Retained Size**
+* **Distance**
```
```suggestion
* **Distance**
* **Shallow Size**
* **Retained Size**
```
same order as in png 
rendered:
https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/memory-problems/memory-101?branch=pr-en-us-3033#object-sizes-and-distances

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,222 +23,175 @@ ms.date: 07/21/2023
+The numbers in the **Shallow size** and **Retained size** columns are the number of bytes.
 
-Learn [how to record a profile using the Heap Profiler](heap-snapshots.md).  In the following figure, some of the notable things in the Heap Snapshot recording in the [Memory](heap-snapshots.md) tool include **Distance**, which is the distance from the garbage-collection root.  If almost all the objects of the same type are at the same distance, and a few are at a bigger distance, that's somethin ...
+![Shallow and Retained size](./memory-101-images/shallow-retained.png)
```
```suggestion
![The Shallow Size and Retained Size columns in the Memory tool](./memory-101-images/shallow-retained.png)
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,222 +23,175 @@ ms.date: 07/21/2023
 
-Node `GC` dominates nodes `#1`, `#3`, and `#11`:
+<!-- ------------------------------ -->
+#### Distance
```
re-order the 3 sections to match png at https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/memory-problems/memory-101?branch=pr-en-us-3033#object-sizes-and-distances 

1. Distance
2. Shallow size
3. Retained size

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,222 +23,175 @@ ms.date: 07/21/2023
-![Node #8 is dominated by node #7 and doesn't dominate any nodes](./memory-101-images/dominators-03.png)
- 
-Node `#10` is dominated by node `#7` and doesn't dominate any nodes:
+An object's retainers are the others objects that hold references to it. The **Retainers** section of the **Memory** tool shows the objects that hold references to the object selected in the **Summary** view.
```
```suggestion
An object's _retainers_ are the others objects that hold references to it. The **Retainers** section of the **Memory** tool shows the objects that hold references to the object selected in the **Summary** view.
```
defining the term 'retainers', so offset via formatting

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,222 +23,175 @@ ms.date: 07/21/2023
+The **Retainers** section of the **Memory** tool is sorted by distance by default, which means that the simplest retaining paths for an object are shown first.
 
-![Node #11 is dominated by node #1 and doesn't dominate any nodes](./memory-101-images/dominators-05.png)
+Any object with no retainers can be discarded by the browser’s garbage collector, which reduces memory usage.
```
```suggestion
Any object with no retainers can be discarded by the browser's garbage collector, which reduces memory usage.
```
straight quotes per Contrib Guide, though skeptical that it makes a difference

GLOBAL nice to do (a separate Comment): change any curly quotes in article to straight quotes.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,222 +23,175 @@ ms.date: 07/21/2023
 *  Immediate 31-bit integer values that are called **small integers** (_SMIs_).
 
-*  Heap objects, referred to as **heap numbers**.  Heap numbers are used for storing values that don't fit into the SMI form, such as **doubles**, or when a value needs to be **boxed**, such as setting properties on it.
+*  Heap objects, referred to as **heap numbers**. Heap numbers are used for storing values that don't fit into the SMI form, such as **doubles**, or when a value needs to be **boxed**, such as setting properties on it.
```
```suggestion
*  Heap objects, referred to as **heap numbers**. Heap numbers are used for storing values that don't fit into the SMI form, such as values of type `double`, or when a value needs to be boxed, such as setting properties on it.
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,222 +23,175 @@ ms.date: 07/21/2023
+#### Other objects
 
-In the following image, in the **Filter edges** dropdown menu, **Hide cycled** is not selected, so a cycled node (grayed out) is displayed:
+* **Native objects**: Anything not in the JavaScript heap is called a _native object_. A native object, in contrast to a heap object, isn't managed by the V8 garbage collector throughout its lifetime, and can only be accessed from JavaScript by using its JavaScript wrapper object.
```
```suggestion
* **Native objects**: Anything that's not stored in the JavaScript heap is called a _native object_. A native object, in contrast to a heap object, isn't managed by the V8 garbage collector throughout its lifetime, and can only be accessed from JavaScript by using its JavaScript wrapper object.
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,222 +23,175 @@ ms.date: 07/21/2023
-![Distance from root](./memory-101-images/root.png)
+The _shallow size_ is the size of the memory that is held by the object directly.
 
+The shallow size of an object is usually small because JavaScript objects often only store the descriptions of the object in memory, and not their values. Arrays and strings are exceptions to this and can have a significant shallow size. Most JavaScript objects store their values elsewhere in the renderer memory, and only expose a small wrapper object on the JavaScript heap.
```
```suggestion
The shallow size of an object is usually small, because JavaScript objects often only store the descriptions of the object in memory, and not their values. Arrays and strings are exceptions to this, and can have a significant shallow size. Most JavaScript objects store their values elsewhere in the renderer memory, and only expose a small wrapper object on the JavaScript heap.
```
commas to help parse.  comma = when reading aloud, a pause.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,222 +23,175 @@ ms.date: 07/21/2023
 
-<!-- ====================================================================== -->
-## Dominators
+Nevertheless, even a small object can hold a large amount of memory indirectly, by preventing other objects from being disposed of by the garbage collection process.
```
```suggestion
Nevertheless, even a small object can hold a large amount of memory _indirectly_, by preventing other objects from being disposed of by the garbage collection process.
```
the intended emphasis

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -24,12 +24,13 @@ ms.date: 07/21/2023
 *  Find memory leaks.
 
-The DevTools heap profiler shows memory distribution used by your page's JavaScript objects and related DOM nodes.  See also [Objects retaining tree](memory-101.md#objects-retaining-tree)) in _Memory terminology_.
+The DevTools heap profiler shows memory distribution used by the JavaScript objects and related DOM nodes on the rendered webpage.
```
```suggestion
The DevTools heap profiler shows the memory distribution used by the JavaScript objects and by related DOM nodes on the rendered webpage.
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -45,136 +46,140 @@ This article uses five demo webpages, all sourced at https://github.com/Microsof
+#### Take another snapshot
 
-![Remove snapshots](./heap-snapshots-images/heap-snapshots-all-hover-clear-all-profiles.png)
+To take another snapshot when one is already displayed in the **Memory** tool, click **Profiles** above the existing snapshot in the sidebar:
```
```suggestion
To take another snapshot when one is already displayed in the **Memory** tool, in the sidebar, click **Profiles** above the existing snapshot:
```
i list steps in order of eye scanning.  First, find X with your eye.  Then find Y in the UI with your eye. ...

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -45,136 +46,140 @@ This article uses five demo webpages, all sourced at https://github.com/Microsof
 ## View snapshots
 
-View snapshots from different perspectives for different tasks.
+Heap snapshots can be viewed in multiple different ways in the **Memory** tool. Each way corresponds to a different task:
```
```suggestion
Heap snapshots can be viewed in multiple different ways in the **Memory** tool. Each way of viewing a heap snapshot in the UI corresponds to a different task:
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -45,136 +46,140 @@ This article uses five demo webpages, all sourced at https://github.com/Microsof
+Heap snapshots can be viewed in multiple different ways in the **Memory** tool. Each way corresponds to a different task:
 
-**Summary view** shows objects grouped by the constructor name.  Use it to hunt down objects (and the memory use) based on type grouped by constructor name.  **Summary view** is particularly helpful for **tracking down DOM leaks**.
+* The **Summary** view shows objects grouped by their constructor name.  Use it to find objects, and the memory they use, based on types grouped by constructor name. The **Summary** view is particularly helpful for **tracking down DOM leaks**.
```
```suggestion
* The **Summary** view shows objects grouped by their constructor name.  Use the *Summary* view to find objects, and the memory they use, based on types grouped by constructor name. The **Summary** view is particularly helpful for tracking down DOM leaks.
```
try to make each sentence self-sufficient & meaningful without relying on "it" pointer indirection that requires resolving "it" by reading a different sentence.

Markdown POV: why bold tracking down DOM leaks?
Rendered POV: bolding might help, across multiple entries, like a table.  Finding: that bolding pattern is NOT used across items in list.

Rendered article section: 
https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots?branch=pr-en-us-3033#view-snapshots

mikehoffms replied:
this list would be more skimmable as a table, to _contrast_/relate the 3 views.  Consider converting list to table.
```
| View | Content | Use for |
|---|---|
| **Summary** | Shows objects grouped by their constructor name. | Finding objects, and the memory they use, based on types that are grouped by constructor name. Helpful for tracking down DOM leaks. |
| **Comparison** | x | x |
| **Containment** | x | x |
```
use this reworded content for row 1

rendered:
https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots?branch=pr-en-us-3033#view-snapshots

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -45,136 +46,140 @@ This article uses five demo webpages, all sourced at https://github.com/Microsof
 <!--todo: add profile memory problems memory diagnosis (tracking down DOM leaks) section when available  -->
 
-**Comparison view**.  Displays the difference between two snapshots.  Use it to compare two (or more) memory snapshots from before and after an operation.  Inspecting the delta in freed memory and reference count lets you confirm the presence and cause of a memory leak.
+* The **Comparison** view displays the differences between two snapshots. Use it to compare two (or more) memory snapshots from before and after an operation.  Inspecting the delta in freed memory and reference count lets you confirm the presence and cause of a memory leak.
```
```suggestion
* The **Comparison** view displays the differences between two snapshots. Use it to compare two (or more) memory snapshots from before and after an operation.  Inspecting the delta in freed memory and inspecting the reference count helps you confirm the presence of a memory leak, and helps determine its cause.
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -45,136 +46,140 @@ This article uses five demo webpages, all sourced at https://github.com/Microsof
 ![Summary view](./heap-snapshots-images/heap-snapshots-constructor-retainers.png)
 
-Top-level entries are "total" lines.
+Top-level entries in the list of constructors are _total lines_.
```
needs rewrite.  check rendered section: 
https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots?branch=pr-en-us-3033#summary-view

are we talking about UI string "total lines"?  or terminology in the industry?

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -45,136 +46,140 @@ This article uses five demo webpages, all sourced at https://github.com/Microsof
 ![Summary view](./heap-snapshots-images/heap-snapshots-constructor-retainers.png)
 
-Top-level entries are "total" lines.
+Top-level entries in the list of constructors are _total lines_.
```
```suggestion
In the **Constructor** column, the top-level entries show a number such as **x123**, indicating the total number of lines affected.
```

mikehoffms replied:
apparently the text is trying to define a term to be reused below, the term _total line_.  Below reads (dubiously): After expanding a total line in the **Summary** view, all of the instances are displayed.

captainbrosset replied:
I'm removing total line and changing this section.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -45,136 +46,140 @@ This article uses five demo webpages, all sourced at https://github.com/Microsof
+| **Retained size** | Displays the maximum retained size among the same set of objects.  The size of memory that you can free after an object is deleted (and the dependents are made no longer reachable) is called the retained size.  See [Retained size](./memory-101.md#retained-size) in _Memory terminology_. |
 
-After expanding a total line in the upper view, all of the instances are displayed.  For each instance, the shallow and retained sizes are displayed in the corresponding columns.  The number after the `@` character is the unique ID of the object, allowing you to compare heap snapshots on per-object basis.
+After expanding a total line in the **Summary** view, all of the instances are displayed.  For each instance, the shallow and retained sizes are displayed in the corresponding columns.  The number after the `@` character is the unique ID of the object, allowing you to compare heap snapshots on per-object basis.
```
rewrite / reformat "a total line", to become grammatical

mikehoffms replied:
```suggestion
After expanding a top-level entry in the **Summary** view, all of the instances are displayed.  For each instance, the shallow and retained sizes are displayed in the corresponding columns.  The number after the `@` character is the unique ID of the object, allowing you to compare heap snapshots on per-object basis.
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -45,136 +46,140 @@ This article uses five demo webpages, all sourced at https://github.com/Microsof
+
+The **Summary** view also contains special category names that aren't based on constructors. These special categories are:
+
+<!-- Taken from https://github.com/sethbrenith/sethbrenith.github.io/blob/main/heap-snapshot-names.md -->
```
```suggestion
<!-- from https://github.com/sethbrenith/sethbrenith.github.io/blob/main/heap-snapshot-names.md -->
```
"taken" has some connotations, not needed

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -45,136 +46,140 @@ This article uses five demo webpages, all sourced at https://github.com/Microsof
-### Heap profiling summary
-todo: add heap profiling summary section when available
--->
+| **(array)** | Various internal array-like objects which don't directly correspond to objects visible from JavaScript, such as the contents of JavaScript Arrays, or the named properties of JavaScript objects. |
```
```suggestion
| **(array)** | Various internal array-like objects that don't directly correspond to objects visible from JavaScript, such as the contents of JavaScript Arrays, or the named properties of JavaScript objects. |
```
i'd probably use "which", but allegedly it should be "that", unless add comma.
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/t/that-vs-which

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -45,136 +46,140 @@ This article uses five demo webpages, all sourced at https://github.com/Microsof
-1. Take a second heap snapshot and change the view of this one to **Comparison**, comparing it to **Snapshot 1**.
+1. In the second heap snapshot, change the view to **Comparison**, comparing it to **Snapshot 1**.
 
 In the **Comparison** view, the difference between two snapshots is displayed.  When expanding a total entry, added and deleted object instances are shown.
```
```suggestion
In the **Comparison** view, the difference between two snapshots is displayed.  When expanding a total entry, added and deleted object instances are shown:
```
rendered:
https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots?branch=pr-en-us-3033#comparison-view

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -45,136 +46,140 @@ This article uses five demo webpages, all sourced at https://github.com/Microsof
+#### Containment view
 
-The **Containment** view is essentially a "bird's eye view" of the objects structure of your application.  It allows you to peek inside function closures, to observe virtual machine (VM) internal objects that together make up your JavaScript objects, and to understand how much memory your application uses at a very low level.
+The **Containment** view is essentially a "bird's eye view" of the objects structure of your application.  It allows you to peek inside function closures, to observe virtual machine (VM) internal objects that together make up your JavaScript objects, and to understand how much memory your application uses at a very low level. <!-- Low level makes little sense since we just said it provided a bird ...
```
at first i thought "the objects structure" was a typo and should be "the object's structure".  try reversing words.
```suggestion
The **Containment** view is essentially a "bird's eye view" of the objects structure of your application.  It allows you to peek inside function closures, to observe virtual machine (VM) internal objects that together make up your JavaScript objects, and to understand how much memory your application uses at a very low level. <!-- "low level" makes little sense, since we just said it provided a "bird's eye view". -->
```
also: per style guide, don't write colloquialism/analogy "bird".  Speak directly/ technically/ literally.

captainbrosset replied:
I'm going to remove the first sentence, it makes no sense to me, and I don't think it brings value.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -263,47 +235,54 @@ To find an object in the collected heap, you can search using **Ctrl+F** and giv
+The heap profiler has the ability to reflect bidirectional dependencies between browser native objects (DOM nodes, CSS rules) and JavaScript objects.  This helps to discover otherwise invisible leaks that happen because of forgotten detached DOM subtrees that remain in memory.
 
-DOM leaks may be bigger than you think.  Consider the following sample.  When is the `#tree` garbage-collected?
+Consider the following code sample.  When is the element with ID `tree` garbage-collected?
```
```suggestion
Consider the following code sample: when is the element with ID `tree` garbage-collected?
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
````diff
@@ -263,47 +235,54 @@ To find an object in the collected heap, you can search using **Ctrl+F** and giv
+Consider the following code sample.  When is the element with ID `tree` garbage-collected?
 
 ```javascript
+// Getting a reference to the #tree element.
````
```suggestion
// Get a reference to the #tree element.
```
GLOBAL: -ing is quite unusual in code comments.  not needed; omit.  some spoken langs allegedly don't have "-ing".

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -263,47 +235,54 @@ To find an object in the collected heap, you can search using **Ctrl+F** and giv
-var body = document.querySelector("body");
 
-body.removeChild(treeRef);
+// Removing the #tree element from the DOM.
```
```suggestion
// Remove the #tree element from the DOM.
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -263,47 +235,54 @@ To find an object in the collected heap, you can search using **Ctrl+F** and giv
+// Getting a reference to the #tree element.
 var treeRef = document.querySelector("#tree");
+
+// Getting a reference to the #leaf element,
```
```suggestion
// Get a reference to the #leaf element,
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
````diff
@@ -263,47 +235,54 @@ To find an object in the collected heap, you can search using **Ctrl+F** and giv
+
+// Now the #tree element can be garbage-collected.
 ```
 
````
above, we introducted the code block by posing a question (a dubious way of writing, but works):

When is the element with ID `tree` garbage-collected?

Here, we must explicitly repeat that phrasing and now deliver the answer in a way that makes it clear the question and the answer to that question.
```suggestion
When is the element with ID `tree` garbage-collected?  The above code indicates that the element with ID `tree` garbage-collected when X.
```
manually apply this change

captainbrosset replied:
Re-worked the entire section to avoid asking the question.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
````diff
@@ -263,47 +235,54 @@ To find an object in the collected heap, you can search using **Ctrl+F** and giv
 ```
 
-The `#leaf` maintains a reference to the relevant parent (parentNode) and recursively up to `#tree`, so only when `leafRef` is nullified is the WHOLE tree under `#tree` a candidate for garbage-collection (GC).
+The element with ID `leaf` maintains a reference to its ancestor nodes in the DOM, which includes the element with ID `tree`. Both `treeRef` and `leafRef` must be nullified for the whole DOM tree under the element with ID `tree` to be garbage-collected.
````
```suggestion
The element with ID `leaf` maintains a reference to its ancestor nodes in the DOM, which includes the element with ID `tree`. Both `treeRef` and `leafRef` must first be nullified, for the whole DOM tree under the element with ID `tree` to be garbage-collected.
```
* first - a "when" question was posed before the code block, so "first" is a suitable when-word.
* comma

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -263,47 +235,54 @@ To find an object in the collected heap, you can search using **Ctrl+F** and giv
-The `#leaf` maintains a reference to the relevant parent (parentNode) and recursively up to `#tree`, so only when `leafRef` is nullified is the WHOLE tree under `#tree` a candidate for garbage-collection (GC).
+The element with ID `leaf` maintains a reference to its ancestor nodes in the DOM, which includes the element with ID `tree`. Both `treeRef` and `leafRef` must be nullified for the whole DOM tree under the element with ID `tree` to be garbage-collected.
 
 ![DOM subtrees](./heap-snapshots-images/memory-problems-tree-gc.png)
```
solution:  move .png up by 1 sentence:

The element with ID leaf maintains a reference to its ancestor nodes in the DOM, which includes the element with ID tree: [colon lead-in]

png

Both treeRef and leafRef must be nullified for the whole DOM tree under the element with ID tree to be garbage-collected.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -263,47 +235,54 @@ To find an object in the collected heap, you can search using **Ctrl+F** and giv
-The `#leaf` maintains a reference to the relevant parent (parentNode) and recursively up to `#tree`, so only when `leafRef` is nullified is the WHOLE tree under `#tree` a candidate for garbage-collection (GC).
+The element with ID `leaf` maintains a reference to its ancestor nodes in the DOM, which includes the element with ID `tree`. Both `treeRef` and `leafRef` must be nullified for the whole DOM tree under the element with ID `tree` to be garbage-collected.
 
 ![DOM subtrees](./heap-snapshots-images/memory-problems-tree-gc.png)
```
important: add red boxes

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -263,47 +235,54 @@ To find an object in the collected heap, you can search using **Ctrl+F** and giv
+<!-- ------------------------------ -->
+#### Demo webpage: Example 6: Leaking DOM nodes
 
 Open the example webpage [Example 6: Leaking DOM nodes](https://microsoftedge.github.io/Demos/devtools-memory-heap-snapshot/example-06.html) in a new window or tab, to understand where DOM nodes might leak, and how to detect such leakage.
```
```suggestion
To understand where DOM nodes might leak, and how to detect such leakage, open the example webpage [Example 6: Leaking DOM nodes](https://microsoftedge.github.io/Demos/devtools-memory-heap-snapshot/example-06.html) in a new window or tab.
```
antipattern (b/c reversed): 
Do A, B, and then C, if you want to D.
not always bad, but consider here.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -317,6 +296,74 @@ To read more about DOM leaks and memory analysis fundamentals, check out [Findin
+
+1. Open the [Eval is evil](https://microsoftedge.github.io/Demos/devtools-memory-heap-snapshot/example-07.html) demo webpage in a new window or tab.
+
+1. Record a first heap snapshot.
```
```suggestion
1. Record a heap snapshot.
```
reword simpler, to avoid my mis-parsing that happened:
Record a snapshot of type "first heap".
Record a first-heap snapshot.  (What's a first-heap snapshot?)

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -317,6 +296,74 @@ To read more about DOM leaks and memory analysis fundamentals, check out [Findin
+
+   In the sidebar, the number below the second snapshot should be larger than the number below the first snapshot. This indicates that more memory is being used by the webpage after clicking the **Closures with eval** button.
+
+1. In the second heap snapshot, change the view to **Comparison**, comparing it to the first one.
```
```suggestion
1. In the second heap snapshot, change the view to **Comparison**, comparing the view or snapshot to the first view or snapshot.
```
manually apply this sketchy wording change

"it" ... "one" - i'm lost.  too many indirection wildcard pointer words.  Speak directly.

the phrase "comparing it to the first one" cannot be made sense of, w/o labor of analysis to resolve indirection pointer words

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -263,47 +235,54 @@ To find an object in the collected heap, you can search using **Ctrl+F** and giv
+<!-- ------------------------------ -->
+#### Demo webpage: Example 9: DOM leaks bigger than expected
 
 Open the example webpage [Example 9: DOM leaks bigger than expected](https://microsoftedge.github.io/Demos/devtools-memory-heap-snapshot/example-09.html) in a new window or tab.
```
```suggestion
To see why a DOM leak might be bigger than expected, open the example webpage [Example 9: DOM leaks bigger than expected](https://microsoftedge.github.io/Demos/devtools-memory-heap-snapshot/example-09.html) in a new window or tab.
```

mikehoffms reviewed (changes requested):
1st-pass review of heap-snapshots.md
3rd-pass review of memory-101.md

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -186,74 +191,41 @@ The **Containment** view is essentially a "bird's eye view" of the objects struc
-Use filters to focus on specific parts of a heap snapshot.  You can filter by node type if, for example, you're only interested in the strings or arrays from the heap.
+Use filters to focus on specific parts of a heap snapshot.
 
 When looking at all the objects in a heap snapshot in the **Memory** tool, it can be difficult to focus on specific objects or retaining paths.  Use the **Node Types** filter when looking at a heap snapshot, to focus on only specific types of nodes.  For example, to see only the arrays and string objects that are in the heap, select the **Array** and **String** entries in the **Node Types** filte ...
```
need to add a colon to form a lead-in gluing text & png.  check rendered:
https://review.learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots?branch=pr-en-us-3033#filter-heap-snapshots-by-node-types

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -186,74 +191,41 @@ The **Containment** view is essentially a "bird's eye view" of the objects struc
 
 When looking at all the objects in a heap snapshot in the **Memory** tool, it can be difficult to focus on specific objects or retaining paths.  Use the **Node Types** filter when looking at a heap snapshot, to focus on only specific types of nodes.  For example, to see only the arrays and string objects that are in the heap, select the **Array** and **String** entries in the **Node Types** filte ...
 
 ![Node Types in a heap snapshot in the Memory tool](heap-snapshots-images/node-types-heap-snapshot.png)
```
png is hard to understand where the "Node Types filter" is on the screen.  Add red box around the **Default v** dropdown?  Even add a text callout in png??

captainbrosset replied:
I'm adding a red box, and I'm mentioning that users need to click **Default** to open the filter.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -186,74 +191,41 @@ The **Containment** view is essentially a "bird's eye view" of the objects struc
-Use filters to focus on specific parts of a heap snapshot.  You can filter by node type if, for example, you're only interested in the strings or arrays from the heap.
+Use filters to focus on specific parts of a heap snapshot.
 
 When looking at all the objects in a heap snapshot in the **Memory** tool, it can be difficult to focus on specific objects or retaining paths.  Use the **Node Types** filter when looking at a heap snapshot, to focus on only specific types of nodes.  For example, to see only the arrays and string objects that are in the heap, select the **Array** and **String** entries in the **Node Types** filte ...
```
```suggestion
When looking at all the objects in a heap snapshot in the **Memory** tool, it can be difficult to focus on specific objects or retaining paths.  When looking at a heap snapshot, to focus on only specific types of nodes, use the **Node Types** filter, in the upper right.  For example, to see only the arrays and string objects in the heap, select the **Array** and **String** entries in the **Node Types** filter:
```

mikehoffms replied:
does tooltip say **Node Types**?
ans: __

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
```diff
@@ -317,6 +296,74 @@ To read more about DOM leaks and memory analysis fundamentals, check out [Findin
+
+   The **Comparison** view shows that new strings were created in the second heap snapshot:
+
+   ![The comparison view, showing that new strings were created in the second snapshot](./heap-snapshots-images/closure-retained-strings.png)
```
```suggestion
   ![The Comparison view, showing that new strings were created in the second snapshot](./heap-snapshots-images/closure-retained-strings.png)
```

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshots.md:
````diff
@@ -317,6 +296,74 @@ To read more about DOM leaks and memory analysis fundamentals, check out [Findin
+}
+```
+
+The following example names the function, which makes it easier to distinguish between closures in the recorder heap snapshot:
````
```suggestion
The following example names the function, which makes it easier to distinguish between closures in the heap snapshot:
```
recorder was confusing to me, seems unnecessary

mikehoffms reviewed (approved):
Approved PR, per Writer/Editor reviews/revisions.

mikehoffms commented on microsoft-edge/devtools-guide-chromium/memory-problems/memory-101.md:
```diff
@@ -23,222 +23,175 @@ ms.date: 07/21/2023
+* **Retained Size**
 
-*  _Edges_ are labelled using the names of _properties_.
+![The Distance, Shallow Size, and Retained Size columns in the Memory tool](./memory-101-images/shallow-retained.png)
```
Add red box on Distance column, like the other two discussed.

captainbrosset replied:
There's already a red box around the Distance column header.
